poethepoet/helpers.py

```diff
@@ -1,5 +1,128 @@
+from enum import Enum
 import re
+from typing import Dict, Optional
 
 
 def is_valid_env_var(var_name: str) -> bool:
-    return bool(re.match("[a-zA-Z_][a-zA-Z0-9_]*", var_name))
+    return bool(re.match("^[a-zA-Z_][a-zA-Z0-9_]*$", var_name))
+
+
+class ParserState(Enum):
+    # Scanning for a new assignment
+    SCAN_VAR_NAME = 0
+    # In a value with no quoting
+    SCAN_VALUE = 1
+    # Inside single quotes
+    IN_SINGLE_QUOTE = 2
+    # Inside double quotes
+    IN_DOUBLE_QUOTE = 3
+
+
+def parse_env_file(content: str) -> Dict[str, str]:
+    """
+    Parses variable assignments from the given string. Expects a subset of bash syntax.
+    """
+    content = content + "\n"
+    result = {}
+    cursor = 0
+    state = ParserState.SCAN_VAR_NAME
+    var_name: Optional[str] = ""
+    var_content = []
+
+    while cursor < len(content):
+        if state == ParserState.SCAN_VAR_NAME:
+            # scan for new variable assignment
+            match = re.search(
+                r"^\s*([a-zA-Z_][a-zA-Z_0-9]*)=", content[cursor:], re.MULTILINE
+            )
+            if match is None:
+                comment_match = re.match(r"^\s*\#.*?\n", content[cursor:])
+                if comment_match:
+                    cursor += comment_match.span()[1]
+                    continue
+                elif re.match(r"^\s*$", content[cursor:], re.MULTILINE):
+                    # The rest of the input is whitespace
+                    break
+                else:
+                    raise ValueError(f"No assignment found at position {cursor}")
+
+            var_name = match.group(1)
+            cursor += match.span()[1]
+            state = ParserState.SCAN_VALUE
+
+        if state == ParserState.SCAN_VALUE:
+            # collect up until the first quote, whitespace, or group of backslashes
+            match = re.search(
+                r"^(.*?)(?:(\s|'|\"|\\+))", content[cursor:], re.MULTILINE
+            )
+            assert match
+            new_var_content, match_terminator = match.groups()
+            var_content.append(new_var_content)
+            cursor += len(new_var_content)
+
+            if match_terminator.isspace():
+                assert var_name
+                result[var_name] = "".join(var_content)
+                var_name = None
+                var_content = []
+                state = ParserState.SCAN_VAR_NAME
+                continue
+
+            if match_terminator == "'":
+                cursor += 1
+                state = ParserState.IN_SINGLE_QUOTE
+
+            elif match_terminator == '"':
+                cursor += 1
+                state = ParserState.IN_DOUBLE_QUOTE
+                continue
+
+            else:
+                # We found one or more backslashes
+                num_backslashes = len(match_terminator)
```

Comment:
I see what you're doing. But won't the regex
```regex
((?:".*?(?<!\\)(\\\\)*?")|(?:'.*?(?<!\\)(\\\\)*?'))
```
Be sufficient?

### Explanation

 - 2 groups: one for handling `"` and one for handling `'`.
 - The middle part (`.*?`) should be self-explanatory
 - The last part (`(?<!\\)(\\\\)*?`) will match an even number of `\` that are not escaped.

Reply:
Hmm, it seemed simpler this way, but that's worth exploring. Note that backslashes don't works as escapes within single quotes, and we need remove active quotes from the result but preserve escaped backslashes as in lines 84 & 120.